### DISCUSSION TITLE
Windows build-in curl  v8 fails to validate TLS cert by IP SAN, Linux curl v7 works as expected

### DISCUSSION THREAD

**Discussion by elvinasp** [curl/curl#15139] [Q&A] [answered]
Hello,

I was making some networking labs and one of the examples was to connect via curl using various hosntname combinations and seeing how SNI behaves. When I connected to demo server IP address using windows build-in curl, it failed to validate certificate. Linux curl from WSL environment on same machine successfully connects to service via IP and validated TLS cert. 

Has anything changed since v7 to v8?

Attaching certs generated with cfssl: [certs.zip](https://github.com/user-attachments/files/17242012/certs.zip)

Windows curl: 
`curl 8.7.1 (Windows) libcurl/8.7.1 Schannel zlib/1.3 WinIDN
Release-Date: 2024-03-27
Protocols: dict file ftp ftps http https imap imaps ipfs ipns mqtt pop3 pop3s smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM SPNEGO SSL SSPI threadsafe Unicode UnixSockets`

Output:
```
* schannel: added 2 certificate(s) from CA file './combined-ca.crt'
* schannel: connection hostname (192.168.4.2) did not match against certificate name (default.tp13.internal)
* schannel: connection hostname (192.168.4.2) did not match against certificate name (default)
* schannel: CertGetNameString() failed to match connection hostname (192.168.4.2) against server certificate names
* Closing connection
* schannel: shutting down SSL/TLS connection with 192.168.4.2 port 8443
curl: (60) schannel: CertGetNameString() failed to match connection hostname (192.168.4.2) against server certificate names
```

Linux curl:
`curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd` 

Output:
```
* Server certificate:
*  subject: C=LT; ST=Kauno raj.; L=Kaunas; O=Centric; OU=DevOps Academy; CN=default.tp13.internal
*  start date: Oct  3 08:14:00 2024 GMT
*  expire date: Oct  3 08:14:00 2025 GMT
*  subjectAltName: host "192.168.4.2" matched cert's IP address!
*  issuer: C=LT; ST=Kauno raj.; L=Kauna; O=Centric; OU=DevOps Academy; CN=DevOps Academy Intermediate CA
*  SSL certificate verify ok.
```

TLS certificate:

Properly lists all SANS:
```
            X509v3 Subject Alternative Name:
                DNS:default.tp13.internal, DNS:default, IP Address:192.168.4.3, IP Address:192.168.4.2, IP Address:10.20.3.1
```

**Comment by bagder** [comment]
> Has anything changed since v7 to v8?

Yes.

~~~
$ git log --oneline curl-7_81_0..curl-8_7_1 -- lib/vtls/schannel.c | wc -l
54

**Reply by elvinasp** [reply]
> git log --oneline curl-7_81_0..curl-8_7_1 -- lib/vtls/schannel.c | wc -l

Ok. Let me rephrase. Has anything changed regarding TLS IP SAN validation? ;)

**Reply by bagder** [reply, marked as answer]
Well, clearly something changed as otherwise I cannot see how the behavior would be different.

This looks like a bug to me. I propose a debug session to try to figure out why/where it goes wrong. The schannel backend is Windows-specific and I don't personally develop on Windows so I have a hard time to reproduce and work on this.

**Reply by elvinasp** [reply]
So, I am not wrong that this seems to be valid to be "escalated" to an issue? As I would expect curl to recognize TLS certificate with IP SAN.

**Comment by icing** [comment]
I think your 7.81.0 version uses OpenSSL. The log output shown is from the OpenSSL backend.

**Comment by bagder** [comment]
It is probably related to this:

> * schannel: added 2 certificate(s) from CA file './combined-ca.crt'

... because I tried `curl.exe` 8.7.1 on a Windows 10 machine next to me and it could fetch `https://1.1.1.1` just fine.

**Comment by jay** [comment]
If a CA certificate bundle is provided and Schannel is the backend then curl does the hostname verification instead of Schannel. In schannel_verify.c [Curl_verify_host](https://github.com/curl/curl/blob/curl-8_10_1/lib/vtls/schannel_verify.c#L470) calls [cert_name_getstring](https://github.com/curl/curl/blob/curl-8_10_1/lib/vtls/schannel_verify.c#L336) which gathers the names. 

If you built your libcurl to target Windows 7 then it manually collects dnsnames [CERT_ALT_NAME_DNS_NAME](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_alt_name_entry) so IP addresses are skipped.

https://github.com/curl/curl/blob/7eb8c048470ed2cc14dca75be9c1cdae7ac8498b/lib/vtls/schannel_verify.c#L436-L438

For Windows 8+ it leaves it to Windows to gather dnsnames which I assume does the same thing

https://github.com/curl/curl/blob/7eb8c048470ed2cc14dca75be9c1cdae7ac8498b/lib/vtls/schannel_verify.c#L365-L381

I'm not sure why there's not support for IP addresses but probably nobody asked for it in Windows

**Reply by elvinasp** [reply]
>If you built your libcurl to support Windows 7, ie you used an old sdk, t

I was using same CA bundle for both Linux and Windows versions to access hosts. Cannot say what SDK was used to build curl, as I was using build-in Windows curl.exe from Windows 10. 

>I'm not sure why there's not support for IP addresses but probably nobody asked for it in Windows

Isn't it an IP support for TLS communication a part of standard TLS implementation? As an end user I would expect so - IP based TLS connectivity is quite common in various network systems. 

It is not a critical issue for my lab exercises, I can add -k option for Windows examples. It just took me a half an hour to understand what is happening and initially I was checking wrong i.e. server side. :) 

I have just tried to download and use 8.7.1 version from GitHub releases for Linux and it works as expected:

```
~/tmp/nginx-server/ssl $ ./curl --ssl-no-revoke --cacert ./combined-ca.crt --header "Host: default.tp13.internal"  https://192.168.4.2:8443 -v
*   Trying 192.168.4.2:8443...
* Connected to 192.168.4.2 (192.168.4.2) port 8443
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: ./combined-ca.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / x25519 / RSASSA-PSS
* ALPN: server accepted h2
* Server certificate:
*  subject: C=LT; ST=Kauno raj.; L=Kaunas; O=Centric; OU=DevOps Academy; CN=default.tp13.internal
*  start date: Oct  3 08:14:00 2024 GMT
*  expire date: Oct  3 08:14:00 2025 GMT
*  subjectAltName: host "192.168.4.2" matched cert's IP address!
*  issuer: C=LT; ST=Kauno raj.; L=Kauna; O=Centric; OU=DevOps Academy; CN=DevOps Academy Intermediate CA
*  SSL certificate verify ok.
*   Certificate level 0: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://192.168.4.2:8443/
```
Will try to download same release for Windows from Github tomorrow, when I power up work laptop - too lazy to do that this evening.

**Reply by bagder** [reply]
> I have just tried to download and use 8.7.1 version from GitHub releases for Linux and it works as expected:

Yes. Your issue is only in the Schannel code and only happens when you use a custom cacert file.

The problem is that *this code* does not handle `CERT_ALT_NAME_IP_ADDRESS`.

https://github.com/curl/curl/blob/bc055d08acb29a8fe05f432640d01fe51379428c/lib/vtls/schannel_verify.c#L436

**Reply by elvinasp** [reply]
Well, it looks like Windows  build of curl 8.7.1 provided in [Curl Download page](https://curl.se/download.html#Win64) also does not have such issue. It behave as expected and properly validates TLS certificate, when connecting to an IP address. 

```
PS C:\Users\epilipon\Downloads\nginx\nginx-server\ssl> ../curl.exe --ssl-no-revoke --cacert ./combined-ca.crt   https://192.168.4.2:8443 -v
*   Trying 192.168.4.2:8443...
* Connected to 192.168.4.2 (192.168.4.2) port 8443
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: ./combined-ca.crt
*  CApath: none
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=LT; ST=Kauno raj.; L=Kaunas; O=Centric; OU=DevOps Academy; CN=default.tp13.internal
*  start date: Oct  3 08:14:00 2024 GMT
*  expire date: Oct  3 08:14:00 2025 GMT
*  subjectAltName: host "192.168.4.2" matched cert's IP address!
*  issuer: C=LT; ST=Kauno raj.; L=Kauna; O=Centric; OU=DevOps Academy; CN=DevOps Academy Intermediate CA
*  SSL certificate verify ok.
*   Certificate level 0: Public key type ? (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 1: Public key type ? (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
* old SSL session ID is stale, removing
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://192.168.4.2:8443/
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: 192.168.4.2:8443]
* [HTTP/2] [1] [:path: /]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
> GET / HTTP/2
....
```

Again it is just a some broken build in Windows internal utility set. Doesn't  that sound familiar? :)

Main question is this something that supposed to be fixed by Curl project or should MS build curl some OTHER way. :)

**Reply by jay** [reply]
It's a curl issue not MS. We'd have to add support for verifying IP addresses when a CA certificate bundle is used with Schannel.

**Reply by elvinasp** [reply]
> It's a curl issue not MS. We'd have to add support for verifying IP addresses when a CA certificate bundle is used with Schannel.

So, should I raise a ticket for this?  At the moment I have adjusted lab with instructions download curl build from GitHub.

**Reply by bagder** [reply]
I already did that: #15149

But this is an area of code without a lot of developer interest so don't get your hopes up for a quick fix.

**Reply by elvinasp** [reply]
> I already did that: #15149
> 
> But this is an area of code without a lot of developer interest so don't get your hopes up for a quick fix.

Well, at least I now know so I'm prepared. :)
